kernel/src/filesystem/vfs/file.rs

```diff
@@ -191,6 +191,256 @@ impl PageCache {
+                }),
+                PageFlags::PG_LRU,
+                &mut LockedFrameAllocator,
+            )?;
+
+            let mut page_guard = page.write_irqsave();
+            unsafe {
+                page_guard.copy_from_slice(&buf[buf_offset..buf_offset + MMArch::PAGE_SIZE]);
+            }
+
+            guard.insert(page_index, page.clone());
+        }
+
+        Ok(())
+    }
+
+    /// 从PageCache中读取数据。
+    ///
+    /// ## 参数
+    ///
+    /// - `offset` 偏移量
+    /// - `buf` 缓冲区
+    ///
+    /// ## 返回值
+    ///
+    /// - `Ok(usize)` 成功读取的长度
+    /// - `Err(SystemError)` 失败返回错误码
+    pub fn read(&self, offset: usize, buf: &mut [u8]) -> Result<usize, SystemError> {
+        let inode = self.inode.upgrade().unwrap();
+        let file_size = inode.metadata().unwrap().size;
+
+        let len = if offset < file_size as usize {
+            core::cmp::min(file_size as usize, offset + buf.len()) - offset
+        } else {
+            0
+        };
+
+        if len == 0 {
+            return Ok(0);
+        }
+
+        let mut not_exist = Vec::new();
+
+        let start_page_index = offset >> MMArch::PAGE_SHIFT;
+        let page_num = (page_align_up(offset + len) >> MMArch::PAGE_SHIFT) - start_page_index;
+
+        let guard = self.pages.lock_irqsave();
+
+        let mut buf_offset = 0;
+        let mut ret = 0;
+        for i in 0..page_num {
+            let page_index = start_page_index + i;
+
+            // 第一个页可能需要计算页内偏移
+            let page_offset = if i == 0 {
+                offset % MMArch::PAGE_SIZE
+            } else {
+                0
+            };
+
+            // 第一个页和最后一个页可能不满
+            let sub_len = if i == 0 {
+                min(len, MMArch::PAGE_SIZE - page_offset)
+            } else if i == page_num - 1 {
+                (offset + len - 1) % MMArch::PAGE_SIZE + 1
+            } else {
+                MMArch::PAGE_SIZE
+            };
+
+            if let Some(page) = guard.get(&page_index) {
+                let sub_buf = &mut buf[buf_offset..(buf_offset + sub_len)];
+                unsafe {
+                    sub_buf.copy_from_slice(
+                        &page.read_irqsave().as_slice()[page_offset..page_offset + sub_len],
+                    );
+                }
+                ret += sub_len;
+            } else if let Some((index, count)) = not_exist.last_mut() {
+                if *index + *count == page_index {
+                    *count += 1;
+                } else {
+                    not_exist.push((page_index, 1));
+                }
+            } else {
+                not_exist.push((page_index, 1));
+            }
+
+            buf_offset += sub_len;
+        }
+
+        drop(guard);
+
+        for (page_index, count) in not_exist {
+            let mut page_buf = vec![0u8; MMArch::PAGE_SIZE * count];
+            inode.read_sync(page_index * MMArch::PAGE_SIZE, page_buf.as_mut())?;
+
+            self.create_pages(page_index, page_buf.as_mut())?;
```

Comment:
这里可以优化掉，
```suggestion
            self.create_pages(page_index, page_buf.as_mut())?;
```
pagecache的read这里耗了2倍内存。还有一次内存拷贝。
可以直接优化成先创建页面，然后拷贝到页内。